tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +224,296 @@ static bool HasCanonicalDimensionNumbers(
+    // einsum_config = "lhs_loop,rhs_loop->out_loop"
+    std::size_t pos_arrow = einsum_config.find(kArrow);
+    std::size_t pos_comma = einsum_config.find(kComma);
+
+    StringRef lhs_loop = einsum_config.substr(0, pos_comma);
+    StringRef rhs_loop = einsum_config.substr(
+        pos_comma + kComma.size(), pos_arrow - (pos_comma + kComma.size()));
+    StringRef out_loop = einsum_config.substr(pos_arrow + kArrow.size());
+
+    // Check for Invalid Configs
+    // 1.Check that there is only maximum 2 inputs
+    // 2.Check that there is only maximum 1 output
+    // 3.Check that there is 1 kArrow
+    if (rhs_loop.find(kComma) != std::string::npos ||
+        out_loop.find(kComma) != std::string::npos ||
+        out_loop.find(kArrow) != std::string::npos) {
+      return rewriter.notifyMatchFailure(op, "Invalid Einsum Config!");
+    }
+
+    // Find result type, if on tensors.
+    ShapedType result_ty;
+    result_ty = this->typeConverter->convertType(op->getResultTypes().front())
+                    .template dyn_cast<RankedTensorType>();
+
+    // Check result type compatibility.
+    if (!result_ty || !(result_ty.getElementType().isSignlessIntOrFloat() ||
+                        result_ty.getElementType().isa<ComplexType>())) {
+      return rewriter.notifyMatchFailure(
+          op, "mismatched operand/result types or iterator count");
+    }
+
+    // Convert the representation to vector<string>
+    SmallVector<std::string> lhs_ein =
+        GetEinsumConfigAsVector(lhs_loop, get_rank(adaptor.lhs()));
+    SmallVector<std::string> rhs_ein =
+        GetEinsumConfigAsVector(rhs_loop, get_rank(adaptor.rhs()));
+    SmallVector<std::string> out_ein =
+        GetEinsumConfigAsVector(out_loop, result_ty.getRank());
+    SmallVector<SmallVector<std::string>> all_loop_dim = {lhs_ein, rhs_ein,
+                                                          out_ein};
+
+    // Check that all operand's batch has same rank
+    if (!CheckBatchHasEqualRank(lhs_ein.size(), lhs_loop, rhs_ein.size(),
+                                rhs_loop, out_ein.size(), out_loop)) {
+      return rewriter.notifyMatchFailure(
+          op, "Invalid Elipsis('...') within Einsum config!");
+    }
+
+    // Find all unique indices in the input and output
+    llvm::SmallSetVector<StringRef, 4> input_ind;
+    llvm::SmallSetVector<StringRef, 4> output_ind;
+    for (const auto& loops_it : llvm::enumerate(all_loop_dim)) {
+      for (const StringRef& dim : loops_it.value()) {
+        if (loops_it.index() < args.size()) {
+          input_ind.insert(dim);
+        } else {
+          output_ind.insert(dim);
+        }
+      }
+    }
+
+    // Check for contraction/summation indices
+    llvm::SmallSetVector<StringRef, 4> reduction_axe =
+        FindSummationAxes(input_ind, output_ind);
+
+    // Find input/output values and types.
+    auto loc = op.getLoc();
+
+    // Setting up Initial buffer for Output Tensor
```

Comment:
I think "buffer" is not accurate. We don't have buffer semantic in Linalg on tensors. We can either delete the comment or say "Prepare init tensor for linalg.generic op".

Reply:
done